byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs

```diff
@@ -0,0 +1,170 @@
+    outs = []
+    wits = []
+    tx = TxWits (Tx [in0, in1] outs) wits
+  in (typeReps tx) @?= typeOf (undefined::TxWits)
+                       <| typeOf (undefined::Tx)
+                       <| typeOf (undefined::[TxIn])
+                       <| typeReps in0
+                       >< typeReps in1
+                       >< (Seq.fromList [ typeOf (undefined::[TxOut])
+                                        , typeOf (undefined::[Wit]) ])
+
+--------------------------------------------------------------------------------
+-- Properties of abstractSize of TxWits / TxIn /TxOut / Wit
+--------------------------------------------------------------------------------
+
+-- | Since 'VKey' appears in both Wit and TxOut (as part of Addr), we
+--   need to use the same "cost" for this type in the following tests.
+vKeyCost :: Size
+vKeyCost = 17
+
+-- | Example costs for types in a 'TxIn'.
+costsTxIn :: Map TypeRep Size
+costsTxIn = Map.fromList [ (typeOf (undefined :: TxIn), 2)
+                         , (typeOf (undefined :: TxId), 3)
+                         , (typeOf (undefined :: Hash), 5) ]
+
+-- | Example costs for types in a 'TxOut'.
+costsTxOut :: Map TypeRep Size
+costsTxOut = Map.fromList [ (typeOf (undefined :: TxOut), 11)
+                          , (typeOf (undefined :: Addr), 13)
+                          , (typeOf (undefined :: VKey), vKeyCost)
+                          , (typeOf (undefined :: Lovelace), 23) ]
+
+-- | Example costs for types in a 'Wit'.
+costsWit :: Map TypeRep Size
+costsWit = Map.fromList [ (typeOf (undefined :: Wit), 31)
+                        , (typeOf (undefined :: VKey), vKeyCost)
+                        , (typeOf (undefined :: Sig Tx), 41)]
+
+-- | This property tests that 'abstractSize' accounts for multiple occurences
+--   of any type represented in the given "cost map".
+--
+--   'abstractSize' uses a "cost map" to compute the size of a term.
+--   For types that contain multiple occurrences of other types, 'abstractSize'
+--   should count each occurence (e.g. a 'Tx' may contain multiple 'TxIn' and
+--   when computing the size of Tx, we should account for the size of each
+--   'TxIn' occurence).
+--
+--  Precondition: this test assumes that the "cost map" only accounts for types
+--  that occur within the "inner" type of multiples that we are targeting, e.g.
+--  if we include a type in the cost map that occurs in 'TxIn' _and_ 'Tx',
+--  this property would have to be weakened to
+--  abstractSize costs x - n * unitCost > 0
+propMultiplesOfSize
+  :: (MonadTest m, HasTypeReps a)
+  => Map TypeRep Size
+  -- ^ a "cost map" for some type that may occur multiple times in the
+  --   given term of type 'a'
+  -> a
+  -- ^ term for which we will compute 'abstractSize'
+  -> Int
+  -- ^ the number of occurences of the "costed types" in given the term
+  -> m ()
+propMultiplesOfSize costs x n
+  = abstractSize costs x === n * unitCost
+  where
+    unitCost = sum (Map.elems costs)
```

Comment:
It's not clear to my why you compute the `unitCost` like this. Won't you be adding all the keys in the map. Is there an implicit assumption that this map contains only the value for the type we're interested in? If this is the case, we might want to make this more explicit by passing a tuple with the type and the cost, and constructing the map in this function.